src/mediaStreams.ts

```diff
@@ -0,0 +1,395 @@
+            } else if (this.session && this.session.onMediaConnectionStateChange) {
+                this.session.onMediaConnectionStateChange(this.session, eventState);
+            } else {
+                this.session.emit('mediaConnectionStateChanged', eventState);
+            }
+        } else {
+            this.rcWPLogd(
+                this.tag,
+                `Unknown peerConnection state: ${sessionDescriptionHandler.peerConnection.iceConnectionState}`
+            );
+        }
+        this.rcWPLogd(this.tag, `peerConnection State: ${eventState}`);
+    }
+
+    public reconnectMedia(options) {
+        let self = this;
+        return new Promise(function(resolve, reject) {
+            if (self.session) {
+                const RTCOptions = {
+                    offerToReceiveAudio: 1,
+                    offerToReceiveVideo: 0,
+                    iceRestart: true
+                };
+                let offerOptions = (options && options.RTCOptions) || RTCOptions;
+                if (!options) {
+                    options = {};
+                }
+                if (!options.extraHeaders) {
+                    options.extraHeaders = self.session.ua.defaultHeaders;
+                }
+                options.eventHandlers = {
+                    succeeded: resolve,
+                    failed: reject
+                };
+                let pc = self.session.sessionDescriptionHandler.peerConnection;
+                pc.createOffer(offerOptions).then(
+                    offer => {
+                        self.session.sessionDescriptionHandler.on('iceCandidate', self.onIceCandidate);
+                        pc.setLocalDescription(offer).then(
+                            () => {
+                                self.rcWPLogd(self.tag, 'reconnecting media');
+                                resolve('reconnecting media');
+                            },
+                            error => {
+                                self.rcWPLoge(self.tag, error);
+                                reject(error);
+                            }
+                        );
+                    },
+                    error => {
+                        self.rcWPLoge(self.tag, error);
+                        reject(error);
+                    }
+                );
+            } else {
+                self.rcWPLoge(self.tag, 'The session cannot be empty');
+            }
+        });
+    }
+
+    private onIceCandidate = (event: any) => {
+        if (event.candidate === null) {
+            this.rcWPLogd(this.tag, 'ice candidate completed for reconnect media');
+            this.session.sessionDescriptionHandler.off('iceCandidate', this.onIceCandidate);
+            this.session.reinvite();
+        }
+    };
+
+    public getRTPReport(reports) {
+        let self = this;
+        let pc = self.session.sessionDescriptionHandler.peerConnection;
+        pc.getStats()
```

Comment:
> We've used https://github.com/muaz-khan/getStats as a safer alternative
> The getStats from github seems have some issue with firefox.

Can you elaborate what kind of issues it has? @vyshakhbabji have you noticed anything like that for QoS in FF?

Reply:
QoS in FF has lot of issues when Elias and I were working on it. That was the reason we donot currently support QoS on FF. Once we move to promise based getStats() which is going to be standardized, we can use getStats() without depending on the above library. I have already finished POC on it and it works. However, there are certain params missing from legacy getStats - I have already spoken to the Analytics team about this - here is the JIRA which has all information required- [here](https://jira.ringcentral.com/browse/SDK-173)

Reply:
QoS in FF has lot of issues when Elias and I were working on it. That was the reason we donot currently support QoS on FF. Once we move to promise based getStats() which is going to be standardized, we can use getStats() without depending on the above library. I have already finished POC on it and it works. However, there are certain params missing from legacy getStats - I have already spoken to the Analytics team about this